Widgets_CDAT/DictWidget/dict.py

```diff
@@ -0,0 +1,219 @@
+
+        # set text based off if valid_keys has values
+        if self.valid_keys:
+            edit_left = QComboBox()
+            for i in self.valid_keys:
+                edit_left.addItem(i)
+            index = edit_left.findText(str(key))
+            if index != -1:
+                edit_left.setCurrentIndex(index)
+
+            # assign signals
+            edit_left.currentIndexChanged.connect(partial(self.updateLists, edit_left, edit_right))
+
+            self.cBoxes.append(edit_left)
+
+        else:
+            edit_left = QLineEdit()
+            validator = InputChecker(self.row_count)
+            self.validators.append(validator)
+            edit_left.setValidator(validator)
+            validator.correctInput.connect(partial(self.colorValid, edit_left))
+            validator.inputInvalid.connect(partial(self.colorInvalid, edit_left))
+            edit_left.setText(str(key))
+
+            edit_left.editingFinished.connect(partial(self.updateLists, edit_left, edit_right))
+            self.textEdits.append(edit_left)
+
+
+        edit_right.editingFinished.connect(partial(self.updateLists, edit_left, edit_right))
+        edit_left.setMinimumWidth(100)
+        col.addWidget(r_button)
+        col.addWidget(edit_left)
+        col.addWidget(edit_right)
+        self.rows.insertLayout(self.row_count, col)
+        self.row_count += 1
+
+
+    def removeRow(self, section, index):
+            child = section.takeAt(0)
+            c = 0
+            while child:
+                widget = child.widget()
+                if type(widget) == QComboBox:
+                    self.cBoxes.remove(widget)
+
+                if c == 1 and type(widget) != QComboBox:
+                    self.validators.remove(self.validators[index])
+
+                    for i in range(len(self.validators)):
+                        self.validators[i].setIndex(i)
+                    self.textEdits.remove(widget)
+                    self.keys.remove(self.keys[index])
+                    self.values.remove(self.values[index])
+
+                widget.deleteLater()
+                child = section.takeAt(0)
+                c += 1
+            section.deleteLater()
+            self.row_count -= 1
+
+            self.textEdited.emit(dict(zip(self.keys, self.values)))
+
+
+    # set valid keys to be selected
+    # must call setValidKeys before calling setDict
+    def setValidKeys(self, keys):
+        self.valid_keys = keys
+        self.valid_keys.insert(0, "")
+        self.cBoxes = []
+
+    # set inital dictionary values
+    def setDict(self, dictionary):
+
+        if len(self.keys) > 0:
+            row = self.rows.takeAt(0)
+            while row:
+                item = row.widget()
+
+                if type(row) == QHBoxLayout:
+                    self.removeRow(row, 0)
+
+                row = self.rows.takeAt(0)
+
+                if type(item) == QPushButton:
+                    self.rows.addWidget(item)
```

Comment:
So you're removing the row, and then adding it back in? How about you structure it so you don't run into this issue by using two QVBoxLayouts, one that holds self.rows and the push button? Then you can iterate across the children to your heart's content.